File: src/ios/ArtKnowledgeGraph/Modules/Storage/SecureStorageModule.m
#import <React/RCTBridgeModule.h> // React Native 0.70.0+
#import <Foundation/Foundation.h> // iOS 14.0+
#import "SecureStorageModule-Swift.h" // Bridge header for Swift implementation

// Error domain constant
NSString * const kSecureStorageErrorDomain = @"com.artknowledgegraph.securestorage";

// Thread-safe queue for storage operations
static dispatch_queue_t secureStorageQueue;

@interface SecureStorageModule : NSObject <RCTBridgeModule>
@property (nonatomic, strong) SecureStorageModule *swiftModule;
@end

@implementation SecureStorageModule

RCT_EXPORT_MODULE();

+ (void)initialize {
    if (self == [SecureStorageModule class]) {
        // Initialize serial queue with user-initiated QoS
        secureStorageQueue = dispatch_queue_create(
            "com.artknowledgegraph.securestorage.queue",
            dispatch_queue_attr_make_with_qos_class(DISPATCH_QUEUE_SERIAL, QOS_CLASS_USER_INITIATED, 0)
        );
    }
}

- (instancetype)init {
    self = [super init];
    if (self) {
        _swiftModule = [[SecureStorageModule alloc] init];
    }
    return self;
}

+ (BOOL)requiresMainQueueSetup {
    return NO; // Using custom queue for operations
}

RCT_EXPORT_METHOD(saveSecureData:(NSString *)key
                  data:(NSString *)data
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
    
    dispatch_async(secureStorageQueue, ^{
        @autoreleasepool {
            // Validate input parameters
            if (!key || [key length] == 0 || !data || [data length] == 0) {
                NSError *error = [NSError errorWithDomain:kSecureStorageErrorDomain
                                                   code:1001
                                               userInfo:@{NSLocalizedDescriptionKey: @"Key and data must not be empty"}];
                reject(@"1001", @"Invalid input parameters", error);
                return;
            }
            
            // Call Swift implementation
            [self.swiftModule saveSecureData:key
                                      data:data
                                  resolve:^(id result) {
                // Clear sensitive data from memory
                @autoreleasepool {
                    NSMutableData *sensitiveData = [data dataUsingEncoding:NSUTF8StringEncoding].mutableCopy;
                    if (sensitiveData) {
                        memset(sensitiveData.mutableBytes, 0, sensitiveData.length);
                    }
                }
                resolve(result);
            }
                                   reject:^(NSString *code, NSString *message, NSError *error) {
                reject(code, message, error);
            }];
        }
    });
}

RCT_EXPORT_METHOD(getSecureData:(NSString *)key
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
    
    dispatch_async(secureStorageQueue, ^{
        @autoreleasepool {
            // Validate key parameter
            if (!key || [key length] == 0) {
                NSError *error = [NSError errorWithDomain:kSecureStorageErrorDomain
                                                   code:1001
                                               userInfo:@{NSLocalizedDescriptionKey: @"Key must not be empty"}];
                reject(@"1001", @"Invalid key parameter", error);
                return;
            }
            
            // Call Swift implementation
            [self.swiftModule getSecureData:key
                                  resolve:^(NSString *result) {
                if (result) {
                    // Return data while ensuring memory cleanup
                    @autoreleasepool {
                        resolve(result);
                    }
                } else {
                    NSError *error = [NSError errorWithDomain:kSecureStorageErrorDomain
                                                       code:1004
                                                   userInfo:@{NSLocalizedDescriptionKey: @"Data not found"}];
                    reject(@"1004", @"Data not found for key", error);
                }
            }
                                   reject:^(NSString *code, NSString *message, NSError *error) {
                reject(code, message, error);
            }];
        }
    });
}

RCT_EXPORT_METHOD(removeSecureData:(NSString *)key
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
    
    dispatch_async(secureStorageQueue, ^{
        @autoreleasepool {
            // Validate key parameter
            if (!key || [key length] == 0) {
                NSError *error = [NSError errorWithDomain:kSecureStorageErrorDomain
                                                   code:1001
                                               userInfo:@{NSLocalizedDescriptionKey: @"Key must not be empty"}];
                reject(@"1001", @"Invalid key parameter", error);
                return;
            }
            
            // Call Swift implementation
            [self.swiftModule removeSecureData:key
                                     resolve:^(id result) {
                resolve(result);
            }
                                      reject:^(NSString *code, NSString *message, NSError *error) {
                reject(code, message, error);
            }];
        }
    });
}

- (void)dealloc {
    // Cleanup any remaining sensitive data
    @autoreleasepool {
        self.swiftModule = nil;
    }
}

@end

File: src/ios/ArtKnowledgeGraph/Modules/Storage/SecureStorageModule.swift
import Foundation // iOS 14.0+
import React // 0.70.0+

// MARK: - Constants
private let kModuleName = "SecureStorageModule"
private let kErrorDomain = "com.artknowledgegraph.app.storage"
private let kStorageQueueLabel = "com.artknowledgegraph.app.storage.queue"
private let kSecureStorageErrorCodes = [
    "invalidInput": 1001,
    "encryptionFailed": 1002,
    "decryptionFailed": 1003,
    "itemNotFound": 1004,
    "deletionFailed": 1005
]

// MARK: - SecureStorageModule Implementation
@objc
@objcMembers
final class SecureStorageModule: NSObject {
    
    // MARK: - Properties
    private let storageQueue: DispatchQueue
    private let operationLock: NSLock
    private let securityManager: SecurityManager
    
    // MARK: - Initialization
    override init() {
        self.storageQueue = DispatchQueue(
            label: kStorageQueueLabel,
            qos: .userInitiated,
            attributes: [],
            autoreleaseFrequency: .workItem,
            target: nil
        )
        self.operationLock = NSLock()
        self.securityManager = SecurityManager.shared
        
        super.init()
        
        // Register for memory pressure notifications
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(handleMemoryWarning),
            name: UIApplication.didReceiveMemoryWarningNotification,
            object: nil
        )
    }
    
    deinit {
        NotificationCenter.default.removeObserver(self)
    }
    
    // MARK: - React Native Required Methods
    @objc
    static func moduleName() -> String {
        return kModuleName
    }
    
    @objc
    static func requiresMainQueueSetup() -> Bool {
        return false
    }
    
    // MARK: - Public Methods
    @objc
    func saveSecureData(_ key: String,
                       data: String,
                       resolve: @escaping RCTPromiseResolveBlock,
                       reject: @escaping RCTPromiseRejectBlock) {
        guard !key.isEmpty, !data.isEmpty else {
            reject(
                String(kSecureStorageErrorCodes["invalidInput"]!),
                "Key and data must not be empty",
                NSError(domain: kErrorDomain, code: kSecureStorageErrorCodes["invalidInput"]!)
            )
            return
        }
        
        storageQueue.async { [weak self] in
            guard let self = self else { return }
            
            self.operationLock.lock()
            autoreleasepool {
                guard let dataToStore = data.data(using: .utf8) else {
                    self.operationLock.unlock()
                    reject(
                        String(kSecureStorageErrorCodes["encryptionFailed"]!),
                        "Failed to encode data",
                        NSError(domain: kErrorDomain, code: kSecureStorageErrorCodes["encryptionFailed"]!)
                    )
                    return
                }
                
                let result = self.securityManager.storeSecureItem(
                    dataToStore,
                    key: key,
                    accessLevel: .whenUnlockedThisDeviceOnly
                )
                
                switch result {
                case .success:
                    // Sanitize sensitive data from memory
                    self.securityManager.sanitizeMemory()
                    self.operationLock.unlock()
                    resolve(true)
                    
                case .failure(let error):
                    self.operationLock.unlock()
                    reject(
                        String(kSecureStorageErrorCodes["encryptionFailed"]!),
                        error.localizedDescription,
                        error as NSError
                    )
                }
            }
        }
    }
    
    @objc
    func getSecureData(_ key: String,
                      resolve: @escaping RCTPromiseResolveBlock,
                      reject: @escaping RCTPromiseRejectBlock) {
        guard !key.isEmpty else {
            reject(
                String(kSecureStorageErrorCodes["invalidInput"]!),
                "Key must not be empty",
                NSError(domain: kErrorDomain, code: kSecureStorageErrorCodes["invalidInput"]!)
            )
            return
        }
        
        storageQueue.async { [weak self] in
            guard let self = self else { return }
            
            self.operationLock.lock()
            autoreleasepool {
                let result = self.securityManager.retrieveSecureItem(
                    key: key,
                    accessLevel: .whenUnlockedThisDeviceOnly
                )
                
                switch result {
                case .success(let data):
                    guard let data = data else {
                        self.operationLock.unlock()
                        reject(
                            String(kSecureStorageErrorCodes["itemNotFound"]!),
                            "Item not found for key: \(key)",
                            NSError(domain: kErrorDomain, code: kSecureStorageErrorCodes["itemNotFound"]!)
                        )
                        return
                    }
                    
                    guard let stringData = String(data: data, encoding: .utf8) else {
                        self.operationLock.unlock()
                        reject(
                            String(kSecureStorageErrorCodes["decryptionFailed"]!),
                            "Failed to decode data",
                            NSError(domain: kErrorDomain, code: kSecureStorageErrorCodes["decryptionFailed"]!)
                        )
                        return
                    }
                    
                    // Sanitize sensitive data from memory
                    self.securityManager.sanitizeMemory()
                    self.operationLock.unlock()
                    resolve(stringData)
                    
                case .failure(let error):
                    self.operationLock.unlock()
                    reject(
                        String(kSecureStorageErrorCodes["decryptionFailed"]!),
                        error.localizedDescription,
                        error as NSError
                    )
                }
            }
        }
    }
    
    @objc
    func removeSecureData(_ key: String,
                         resolve: @escaping RCTPromiseResolveBlock,
                         reject: @escaping RCTPromiseRejectBlock) {
        guard !key.isEmpty else {
            reject(
                String(kSecureStorageErrorCodes["invalidInput"]!),
                "Key must not be empty",
                NSError(domain: kErrorDomain, code: kSecureStorageErrorCodes["invalidInput"]!)
            )
            return
        }
        
        storageQueue.async { [weak self] in
            guard let self = self else { return }
            
            self.operationLock.lock()
            
            let result = self.securityManager.removeSecureItem(key: key)
            
            switch result {
            case .success:
                self.operationLock.unlock()
                resolve(true)
                
            case .failure(let error):
                self.operationLock.unlock()
                reject(
                    String(kSecureStorageErrorCodes["deletionFailed"]!),
                    error.localizedDescription,
                    error as NSError
                )
            }
        }
    }
    
    // MARK: - Memory Management
    @objc
    private func handleMemoryWarning() {
        storageQueue.async { [weak self] in
            guard let self = self else { return }
            
            self.operationLock.lock()
            autoreleasepool {
                self.securityManager.sanitizeMemory()
            }
            self.operationLock.unlock()
        }
    }
}